### PR TITLE
feat: change image class to use PyTorch tensors and methods

### LINKED ISSUE #505: Change Image Class to use PyTorch tensors & PyTorch methods
### Is your feature request related to a problem?

Currently, the Image Class uses pillow.

### Desired solution

The Image Class should use PyTorch tensors and methods from the PyTorch functional API (v2)

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by Marsmaennchen221**
Closes #505 

### Summary of Changes

feat: Changed Image class to use PyTorch tensors
refactor: Image class now uses PyTorch transform methods

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -15,33 +15,40 @@ packages = [
...
 pillow = ">=9.5,<11.0"
-scikit-image = ">=0.21,<0.23"
 scikit-learn = "^1.2.0"
 seaborn = "^0.13.0"
+openpyxl = "^3.1.2"
+scikit-image = ">=0.21,<0.23"
+levenshtein = ">=0.21.1,<0.24.0"
+torch = {version = "~2.0.0", source = "torch_cuda118"}
+torchvision = {version = "<0.16.0", source = "torch_cuda118"}
+syrupy = "^4.6.0"
```
Comment (lars-reimann) [inline pyproject.toml:28]: This should still be a dev dependency.

File: src/safeds/data/image/containers/_image.py [hunk 2]
```diff
@@ -23,71 +30,106 @@ class Image:
...
+    image_tensor : Tensor
+        The image data as tensor.
     """
 
+    _pil_to_tensor = PILToTensor()
+    _default_device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+
     @staticmethod
-    def from_jpeg_file(path: str | Path) -> Image:
+    def from_file(path: str | Path, device: Device = _default_device) -> Image:
```
Comment (lars-reimann) [inline src/safeds/data/image/containers/_image.py:41]: The `device` parameter leaks implementation details. A novice should not have to know that an image is stored internally as a tensor and that the tensor can be stored on different devices. Remove this parameter, please and call `_get_device` (see above) instead.

File: src/safeds/data/image/containers/_image.py [hunk 3]
```diff
@@ -23,71 +30,106 @@ class Image:
...
     Parameters
     ----------
-    data : BinaryIO
-        The image data as bytes.
+    image_tensor : Tensor
+        The image data as tensor.
     """
 
+    _pil_to_tensor = PILToTensor()
+    _default_device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
```
Comment (lars-reimann) [inline src/safeds/data/image/containers/_image.py:38]: The device is not only relevant for images but also for NNs. Thus, this should be moved to another location. I'd suggest something like the following in `src/safeds/config/_device.py` (+ and entry in the corresponding `__init__.py`):

```py
def _get_device() -> ?:
    return torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
```

We can later think about if, when, and how, the device can be overridden.

File: src/safeds/data/image/containers/_image.py [hunk 4]
```diff
@@ -23,71 +30,106 @@ class Image:
...
         """
-        return Image(
-            data=Path(path).open("rb"),
-            format_=ImageFormat.JPEG,
-        )
+        return Image(image_tensor=Image._pil_to_tensor(pil_image_open(path)), device=device)
 
     @staticmethod
-    def from_png_file(path: str | Path) -> Image:
+    def from_bytes(data: bytes, device: Device = _default_device) -> Image:
```
Comment (lars-reimann) [inline src/safeds/data/image/containers/_image.py:60]: Same issue regarding `device`.

File: src/safeds/data/image/containers/_image.py [hunk 5]
```diff
@@ -23,71 +30,106 @@ class Image:
...
+            warnings.filterwarnings(
+                "ignore",
+                message="The given buffer is not writable, and PyTorch does not support non-writable tensors.",
+            )
+            input_tensor = torch.frombuffer(data, dtype=torch.uint8)
+        return Image(image_tensor=torchvision.io.decode_image(input_tensor), device=device)
 
-    def __init__(self, data: BinaryIO, format_: ImageFormat):
-        data.seek(0)
+    def __init__(self, image_tensor: Tensor, device: Device = _default_device) -> None:
```
Comment (lars-reimann) [inline src/safeds/data/image/containers/_image.py:84]: Same issue regarding device.

File: src/safeds/data/image/containers/_image.py [hunk 6]
```diff
@@ -23,71 +30,106 @@ class Image:
...
+        -------
+        png : bytes
+            The image as PNG.
+        """
+        buffer = io.BytesIO()
+        save_image(self._image_tensor.to(torch.float32) / 255, buffer, format="png")
+        buffer.seek(0)
+        return buffer.read()
+
+    def _set_device(self, device: Device) -> Image:
```
Comment (lars-reimann) [inline src/safeds/data/image/containers/_image.py:119]: Same issue regarding device (remove).

File: src/safeds/data/image/containers/_image.py [hunk 7]
```diff
@@ -111,7 +153,31 @@ def height(self) -> int:
...
+
+        Returns
+        -------
+        channel : int
+            The number of channels of the image.
+        """
+        return self._image_tensor.size(dim=0)
+
+    @property
+    def device(self) -> Device:
```
Comment (lars-reimann) [inline src/safeds/data/image/containers/_image.py:171]: Same issue regarding device (remove).

File: src/safeds/data/image/containers/_image.py [hunk 8]
```diff
@@ -139,13 +207,13 @@ def to_png_file(self, path: str | Path) -> None:
             The path to the PNG file.
         """
         Path(path).parent.mkdir(parents=True, exist_ok=True)
-        self._image.save(path, format="png")
+        save_image(self._image_tensor.to(torch.float32) / 255, path, format="png")
 
     # ------------------------------------------------------------------------------------------------------------------
     # IPython integration
     # ------------------------------------------------------------------------------------------------------------------
```
Comment (lars-reimann) [inline src/safeds/data/image/containers/_image.py:214]: Something is not right here: `_repr_jpeg_` and `_repr_png_` were used for the IPython integration. `__eq__` is just the normal operator overload (was wrong before).

File: src/safeds/data/image/containers/_image.py [hunk 9]
```diff
@@ -506,22 +544,7 @@ def rotate_left(self) -> Image:
...
-        -------
-        result : Image
-            The image with edges found.
-        """
-        image_copy = copy.deepcopy(self)
-        image_copy = image_copy.convert_to_grayscale()
-        image_copy._image = image_copy._image.filter(ImageFilter.FIND_EDGES)
-        return image_copy
+    # def find_edges(self) -> Image:
+    #     pass
```
Comment (lars-reimann) [inline src/safeds/data/image/containers/_image.py:550]: Remove + add an issue. Otherwise, we'll forget about it.

**Review by lars-reimann — CHANGES_REQUESTED** [review]
Looks really comprehensive. I only have a few remarks, mostly regarding the handling of devices.

File: src/safeds/data/image/containers/_image.py (hunk 2 above)
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image.py:41]: see #524

File: src/safeds/data/image/containers/_image.py (hunk 4 above)
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image.py:60]: see #524

File: src/safeds/data/image/containers/_image.py (hunk 6 above)
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image.py:119]: see #524

File: src/safeds/data/image/containers/_image.py (hunk 7 above)
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image.py:171]: see #524

File: src/safeds/data/image/containers/_image.py (hunk 9 above)
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image.py:550]: see #523

File: src/safeds/data/image/containers/_image.py (hunk 5 above)
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image.py:84]: see #524

File: src/safeds/data/image/containers/_image.py (hunk 3 above)
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image.py:38]: see #524

**Review by lars-reimann — APPROVED** [review]
Well done, thank you!

**Pull request merged**